### ISSUE TITLE
pihole-FTL stops answering DNS for a while when using the "show all" option in the query log Web UI, returns broken data

### ISSUE THREAD

**Issue by dermoth** [pi-hole/FTL#1549]
pihole-FTL hangs and remain unresponsive even to cached DNS queries; this is testing directly with curl on the API, `http://pi.hole/admin/api.php?getAllQueries`.

During this time the one of the telnet-IPv4-* thread is CPU-bound- I would expect only that thread blocks and the rest of pihole-FTL could function but it's impossible to resolve anything during this time and the pihole log shows a complete gap of 27s (there is normally not one second without something happening in the log) and FTL log shows a warning that the database is locked around the time of the api call.

The returned data also seems to be broken, unless the FE does some kind of json streaming and doesn't care if the list is not terminated (that is doubtful but I've seen command-line clients doing that):

```text
$ jq -r keys response.json 
parse error: Unfinished JSON term at EOF at line 1, column 5195925
```

This time the answer was just about 5M but before updating to the latest version I'd get ~11M with similar behavior/delay.

The behavior is even worse if I try from the Web UI. I haven't tried to find out what happen on the browser side but after the ~30s hang following `getAllQueries` I get this in the FTL log:

```text
[2023-04-02 14:15:10.783 18240/T18360] WARN: Could not write() everything in getAllQueries() [src/api/api.c:1070]: Connection reset by peer
[2023-04-02 14:15:10.783 18240/T18360] WARN: Could not write() everything in getAllQueries() [src/api/api.c:1074]: Broken pipe
[... 228271 repeats!! ]
[2023-04-02 14:15:36.220 18240/T18360] WARN: Could not write() everything in getAllQueries() [src/api/api.c:1074]: Broken pipe
[2023-04-02 14:15:36.221 18240/T18360] WARN: Could not write() everything in seom() [src/api/socket.c:253]: Broken pipe
```

In at least one instance ( suspect more before I started logging what was happening) FTL appeared to remain unresponsive after that, printing a few "database is locked" messages until I restarted it about 50 seconds later (I didn't check thread usage unfortunately but in an earlier occurrence when I had to restart FTL it was hung at 100% CPU... I would usually not wait very long for recovery as this is a 1-server setup).

### Versions

Current Pi-hole version is v5.16.2
Current AdminLTE version is v5.19
Current FTL version is v5.22

### Platform

RPi-4, Debian 10.13 (fully updated)

### Expected behavior

1. FTL should remain responsive during the query.
2. The response should be valid json, and possibly have a hard-limit if FTL cannot return a very large dataset without hanging.
3. FTL should detect connection resets and stop trying to send all the the remaining chunks (and flooding the log with errors)

### Actual behavior / bug

FTL hands during long queries. and spam the log on connection reset.

### Steps to reproduce

1. On a setup with large query log, open the Query log in the WebUI and click "show all", capture the request as curl in devtools.
2. In parallel, try to resolve other domains, FTL hangs for a while then spams the log if the connection gets reset (this was using Chrome)
3. Repeat the query using curl.
4. In parallel, try to resolve other domains, FTL hangs like before.
5. Try to parse the json (ex with jq), it's broken (appears unfinished)

## Debug Token

Will provide one on demand. I would preferably just upload relevant pihole log portions

### Screenshots

N/A

### Additional context

Home setup with ~450k active queries (the "active queries" in the dashboard is daily right?)

**Comment by DL6ER** [comment]
This points towards a system that is barely (or rather: not) capable of handling the enormous amount of queries. My assumption seems kind of backed by the fact that this is the very first "bug" report we ever receive about this.

In case of an overwhelmed system, this is actually expected. Behind the scenes, FTL works with locking to avoid race-collisions between its various threads (and forks). What you are observing here is that the API gets an exclusive lock on the data while it is reading. DNS appears to be hanging because no new queries can be added at the time of this exclusive lock.

FTL is typically fast in generating the output (even in the 100,000+ queries range), however, it can only output as fast as the consumer can receive. Note that in Pi-hole v5.x the entirety of the data is still going through PHP which is creating the JSON you are seeing. This PHP script is likely running into a timeout causing the truncated/unfinished JSON payload you are seeing.

We are actively developing the next generation Pi-hole (v6.0) behind the scenes (branch `new/http`) which won't be affected by this. We change from exclusive locking on the entirety of the data to actually having two datasets - one is "new queries", the other one "all queries". The first will typically add immediately to the second one. However, when the second dataset is locked (due to a long/complex API request), the first one will just build up and ensure no hanging on the DNS side is happening. Once the long-running API request finished, all these "new" queries will be committed to the second dataset and everything can continue as before. While this was actually intended to address another issue, it will cover your case 100%, too.
Furthermore, the next-gen FTL v6.0 stops doing the detour over PHP for generating JSON and generates JSON on its own, doing away with any additional delays and timeouts.

**Comment by dermoth** [comment]
Thanks for the reply. Yes php is likely not helping here but since it's a multi-core CPU it's still able to do some work in parallel... On the other issues more specifically:

1. Would it be possible not to lock the database  during queries? I have often run large queries directly on the sqlite file without causing any locks.

2. The insane logging of FTL should be fixed - 228271 lines in less than 30 seconds is nonsense.

3. Should FTL have a default cap on the all queries results to avoid issues? Even if it works, the browser will likely hang processing them (and that issue has been reported in the past).

-----

FWIW the sqlite client can dump much faster too... I'm not sure what the FTL thread was doing but just to compare I can dump >500k rows in less than 15s:

```
pi@raspberrypi:~ $ /usr/bin/time sqlite3 /etc/pihole/pihole-FTL.db <<<$'.headers off\n.mode csv\nSELECT * FROM queries q WHERE q.timestamp > strftime(\'%s\', \'now\', \'localtime\') - 172800;' |wc -l
13.36user 0.10system 0:13.46elapsed 99%CPU (0avgtext+0avgdata 5016maxresident)k
0inputs+0outputs (0major+680minor)pagefaults 0swaps
501054
```

Also now that I'm down at ~153k daily queries, I can actually dump successfully all queries using curl, but it took 2m18  and I received 17MB of data - the fact it breaks much faster on larger datasets makes me think we're actually hitting an internal limit in FTL.

**Comment by DL6ER** [comment]
> Would it be possible not to lock the database during queries? I have often run large queries directly on the sqlite file without causing any locks.

We are not locking the database, there is an internal in-memory structure that is used here.

> The insane logging of FTL should be fixed - 228271 lines in less than 30 seconds is nonsense.

What is being logged? When no debug flags are set, there should be zero logging (except warnings and errors).

> Should FTL have a default cap on the all queries results to avoid issues? Even if it works, the browser will likely hang processing them (and that issue has been reported in the past).

Yes, this may be a good idea. As I said before, we've not seen such a report before so it didn't seem necessary so far.

> FWIW the sqlite client can dump much faster too... I'm not sure what the FTL thread was doing but just to compare I can dump >500k rows in less than 15

FTL is not using the database to create the output.

---

Please try to remove PHP from the picture when measuring the CPU time. Running something like
```
time echo ">getallqueries >quit" | nc 127.0.0.1 4711 > /dev/null
```
should be more appropriate.
